### PR TITLE
Heuristic to detect if a gamecube game is rendering 16:9 widescreen.

### PR DESCRIPTION
**Pull request by phire**
Someone suggested on IRC that we should make a database of memory locations in GameCube games which contain the 'Widescreen' setting so we can automatically detect if the game is in 4:3 or 16:9 mode. 

But that's hardly optimal, when the game actually tells the gpu what aspect ratio to render in. 10 min and 6 lines of code later, this is the result. Not only does it detect the correct aspect ratio, it does so on the fly.

I'm a little surprised that nobody thought about doing this before.

Tested with F-Zero GX and Burnout 2.

Burnout 2 doesn't detect the changed aspect ratio until you actually start a race, as the menu is completely orthographic and doesn't even attempt to adjust for the 16:9 aspect ratio, except for the spinning cars, which are a video.

### CODE REVIEW

**Comment by phire** [comment]
Also works with the Widescreen AR code for windwaker, even switches from 16:9 to 4:3 when you pull up the item screen.

**Comment by JMC47** [comment]
Merge it.  This is gold.

**Comment by JMC47** [comment]
Works for Tony Hawk Pro Skater 4; switches to 16:9 upon loading a save file with 16:9 enabled or switching the option.  100% seamless and perfect.  Why isn't this merged yet?  Damn code review.

**Comment by JMC47** [comment]
Tested in more games.  LGTM.

**Comment by degasus** [comment]
Isn't this information provided by VI?

**Comment by delroth** [comment]
What's up with zww-armos on fifoci showing up in 16:9 now?

**Comment by JMC47** [comment]
That's on me; it was originally recorded with the widescreen hack on.

**Comment by phire** [comment]
@degasus No, Anamorphic widescreen doesn't change the configuration of VI at all. The user is required to manually change the aspect ratio on their tv.

**Comment by JMC47** [comment]
Any reason not to merge this after discussion?

**Comment by phire** [comment]
Let me improve/document the magic constants of 1.7, 1.85, 1.25 and 1.4, which I picked kind of arbitrarily.

Then it can be merged.

File: Source/Core/VideoCommon/VertexShaderManager.cpp [hunk 1]
```diff
@@ -417,6 +419,21 @@ void VertexShaderManager::SetConstants()
...
+			// Heuristic to detect if a GameCube game is in 16:9 anamorphic widescreen mode.
+			if (!SConfig::GetInstance().bWii)
+			{
+				// Due to the BT.601 standard which the GameCube is based on being a compromise 
+				// between PAL and NTSC, neither standard gets square pixels. They are each off 
+				// by ~9% in opposite directions.
+				// Just in case any game decides to take this into account, we do these tests
+				// with a large amount of slop.
+				float aspect = fabsf(rawProjection[2] / rawProjection[0]);
+				if (fabsf(aspect - 16.0f/9.0f) < 16.0f/9.0f * 0.11) // within 11% of 16:9
```
Comment (BhaaLseN) [inline Source/Core/VideoCommon/VertexShaderManager.cpp]: I personally prefer spaces around the ```/```; but what do you think of making those two name constants?

**Comment by bdr99** [comment]
It was me who originally suggested this on IRC, but I never envisioned such a clean implementation as this. Thank you phire for implementing my idea. I tested it and it works exactly as I imagined.

**Comment by BFG1992** [comment]
Hello.
I've made a quick testing of this. Launched Wind Waker through Swiss launcher with widescreen hack on in Swiss' options (the reason is because of better Swiss' widescreen hack realization - no UI stretching). The Swiss itself is 4:3. The game had switched to widescreen, but done it in a wierd way:
!(http://ipic.su/img/img7/fs/kiss_419kb.1438357618.png)
This is happening only when Aspect Ratio in Graphics options) is set to Auto before launch, if I manually set it in 16:9 and then to Auto during gameplay, it remains widescreened correctly.
Also to note: Swiss itself is wierd - it doesn't update frame output until any action is done, FPS remains at 0 in render window header. The VPS and speed is 60 and 100% though. Maybe that wierd game behaviour is because of this?

Anyway, this is really nice feature, thank you, devs! :)

**Comment by JMC47** [comment]
Honestly, I don't know what to say.  It looks like Swiss is doing something really weird rather than Dolphin doing something unexpected.  It's likely just following the rules given to it while Swiss is doing some serious hack to get widescreen.

**Comment by BFG1992** [comment]
Done more testing now. It seems that this wierd behaviour changes to normal if I enter or exit fullscreen mode.
It also fixes itself if Real XFB is on (Swiss starts to show 60 fps too). I think maybe Dolphin has some troubles when initializing change of rendering from Swiss' wierd one to Wind Waker's normal one, but after video output reinitialization things becomes normal again (its just my uneducated thoughts though).
Also, that's how it looks if I run it in windowed mode:
ww_windowed

**Comment by sigmabeta** [comment]
This works beautifully for me in F-Zero GX, but I quickly tried out Metroid Prime 2 and saw that it seems to trigger widescreen whenever a 'cutscene' is happening, but does not appear to adjust the aspect ratio / transformation to reflect this:

g2me01-2

To reproduce, start any new game and watch the opening cutscene - or, this also applies to the short cutscene that plays when you save or load a game.

**Comment by sigmabeta** [comment]
Also happens in Mario Kart Double Dash. Game works correctly until the moment you select your first driver, then it stretches to 16:9. The only way to get it to unstretch is to start a race:

gm4e01-1

gm4e01-2

**Comment by BFG1992** [comment]
I can confirm an issue with Metroid Prime 2 too. The output stretches on all cutscenes back to 4:3, and doesn't care for "Force 16:9" setting.
Maybe Dolphin should apply this heuristic technique only on "Auto" aspect selected?

**Comment by JMC47** [comment]
That's exactly what it does?  You can disable it by setting the aspect ratio to 4:3 instead of auto.

**Comment by BFG1992** [comment]
@JMC47 But I want to play in 16:9 with widescreen hack (not Swiss', but Dolphin's native one this time). Before this was merged, anything was fine, but now the thing stretches on cutscenes.

**Comment by JMC47** [comment]
Then select Force 16:9?  Auto is the only one with the heuristic applied.

**Comment by BFG1992** [comment]
@JMC47 But I selected Force 16:9. As I stated earlier:

>The output stretches on all cutscenes back to 4:3 (including even save stations ones), and doesn't care for "Force 16:9" setting.

If you want, I can make a video showing it.

**Comment by JMC47** [comment]
That's gotta be a bug or something, force 16:9/4:3 shouldn't be changed...

**Comment by BFG1992** [comment]
Here's what I mean:
mp2wsbug

**Comment by BFG1992** [comment]
Oh, I found something else :)
When I tested Zelda WW wierd behaviour... there's something similar:
mp2bug
To fix it to 4:3 - just resize window or enter fullscreen.

**Comment by BFG1992** [comment]
I've been thinking about it, and made a video to show this bug more clearly: https://www.youtube.com/watch?v=RQyIEhMtDYE

**Comment by yorkiesandskittles** [comment]
Brawl/Project M stay at 16:9 even when 4:3 is selected in Options.

**Comment by mirrorbender** [comment]
This bug should be fixed once my aspect ratio pull request #2765 is merged.

**Comment by phire** [comment]
Ok, Metroid prime 2 and Double Dash should be fixed in #2790 

The issues with force 4:3 not working should be fixed by mirrorbender's PR, which is now merged.

**Comment by phire** [comment]
With the swiss widescreen hack, you probably need to enable virtual xfb for that to work. I personally recommend the 16:9 AR code from [our wiki page](https://wiki.dolphin-emu.org/index.php?title=The_Legend_of_Zelda:_The_Wind_Waker)

**Comment by BFG1992** [comment]
@phire Swiss works fine with XFB off, but shows zero fps until something on screen is changed (a menu triggered, a selection moved etc.). With Real XFB it works even better - the FPS shows correct 60.
But with Virtual XFB it works really odd - the actions render with one action lag. I dunno how to explain it... if you move the selection in list three times down, it will show that you moved it only two times. But if you next do something else (choose menu item or so), the image will change to what you did before it. It will move the selection cursor to one position down in our case. If you then do yet anoher action, it will now show as if you chose menu item. And so on
I can make a video of this oddness if you want, but Swiss is free and small in size, you can test it easily by yourself, if you interested. Here's the link to the version I'm using: https://yadi.sk/d/P3x2i6U5iD6bo

As for AR code - it strectches the UI. I hate stretched UI, so I use Swiss' widescreen hack.
Also, since Swiss in open source, then why didn't anyone looked into the code behind it? It's more advanced than Dolphin's one. Maybe some will port it?

**Comment by mimimi085181** [comment]
Does Swiss' widescreen hack only keep obejects on screen in the area that is invisible on 4:3? If you don't know what i mean, turn on the regular widescreen hack in Dolphin(not the AR code) and run around the 1st village.

**Comment by JMC47** [comment]
I tried to port swiss a long time ago, but as far as I can tell it's a bunch of game specific hacks and no one who is interested in an implementation wants to do a bunch of per-game hacks.

**Comment by BFG1992** [comment]
@mimimi085181 You mean polygons clipping on the side edges of the widened screen? Yes, just like Dolphin's widehack. These hacks - Swiss' one and Dolphin's one - works just identically, but the Swiss' one doesn't stretch the interface (it will remain in the same place as in 4:3)

**Comment by mimimi085181** [comment]
Then, i still prefer the AR code, but i guess that's personal preference. If you want both, take the AR code and make custom textures for the interface.

If the Swiss' widescreen hack was generic, i'd maybe try to port it. But i can't do much in the next few weeks.

**Comment by BFG1992** [comment]
@JMC47 In Swiss' options there's only two widehacks one can choose: Persp and Yes. What to choose depends of game. For LoZ:WW it's Yes. Maybe you mean there's hacks in the code, not visible to user? Then why don't port the bulk of Swiss' widehack (Persp and Yes options), and leave per-game hacks be? I think it will benefit for many games even this way (or don't - just my uneducated guess).

@mimimi085181
>that's personal preference
This. I can play with clipping, but the looks of streeeeetched interface kinda makes me suffer :grinning:

**Comment by BFG1992** [comment]
Tested games after these merges again. Thanks for work, folks! Metroid Prime 2 now isn't sctretching anything.
But there is something strange with Zelda WW (no Swiss): it looks like 16:10 in Force 16:9 and 5:4 in Force 4:3 or Auto.
Maybe its related to mimimi085181's VI scaling fix?

**Comment by mimimi085181** [comment]
I didn't do anything!

**Comment by BFG1992** [comment]
@mirrorbender Sorry! I confused you with mirrorbender >_<

**Comment by phire** [comment]
Mirrorbender's VI scaling fix.

It will be very slightly (not 16:10) off 16:9 due to the way overscan is handled. If you use the crop option in advanced it will crop it to a 16:9 image.

**Comment by BFG1992** [comment]
Looks like this:
zww_1610

**Comment by BFG1992** [comment]
Launched Mario Strikers Charged PAL:
msc_pal
The menu is not a perfect circle, like it must be. Drawn a perfect one to show.

**Comment by phire** [comment]
Check on a real TV and gamecube. It also won't be a perfect circle.

**Comment by BFG1992** [comment]
@phire Well, maybe, but I think the developers may not designed the game to look this way. That distortion is made by Wii itself, right? Than, though this distortion is a more correct emulation, there must be an option to override it, I think.
P.S.: Sorry if I'm being annoying already...

**Comment by JMC47** [comment]
There's already a PR to override it when you want to https://github.com/dolphin-emu/dolphin/pull/2791

**Comment by BFG1992** [comment]
@JMC47 Thanks! Haven't seen it yet.

**Comment by SMarioMan** [comment]
Thanks for implementing this. Works great on Super Monkey Ball Adventure. I was pleasantly surprised when it happened.

**Pull request merged**